rfcs/text/0017_url_service.md

```diff
@@ -0,0 +1,462 @@
+}
+```
+
+Note, that in the new service to generate a short URL the developer will have to
+use a URL generator.
+
+```ts
+const shortUrl = await share.url.shortUrls.create({
+  generator: 'DISCOVER_DEEP_LINKS',
+  state: {
+    indexPattern: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx',
+  },
+});
+```
+
+These short URL will be stored in saved objects of type `url` and will be
+automatically migrated once a URL generator becomes deprecated.
+
+
+## HTTP endpoints
+
+
+### Short URL CRUD+ HTTP endpoints
+
+Below HTTP endpoints are designed to work specifically with short URLs:
+
+| HTTP method           | Path                                | Description                                                                                                                              |
+|-----------------------|-------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------|
+| __POST__              | `/api/short_url`                    | Endpoint for creating new short URLs.                                                                                                    |
+| __GET__               | `/api/short_url/<slug>`             | Endpoint for retrieving information about an existing short URL.                                                                         |
+| __DELETE__            | `/api/short_url/<slug>`             | Endpoint for deleting an existing short URL.                                                                                             |
+| __POST__              | `/api/short_url/<slug>`             | Endpoint for updating information about an existing short URL.                                                                           |
+| __GET__ or __POST__   | `/api/short_url/<slug>/_resolve`    | Same as `GET /api/short_url/<slug>`, but this endpoint also increments the short URL access count counter and the last access timestamp. |
+| __GET__ or __POST__   | `/api/short_url/<slug>/_redirect`   | Redirects the user using HTTP 301 (permanent) redirect to the destination long URL.                                                      |
+
+
+
+### Other URL Service HTTP endpoints
+
+Below are all the other HTTP endpoints that URL Service will provide:
+
+| HTTP method           | Path                                | Description                                                                                                                              |
+|-----------------------|-------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------|
+| __GET__ or __POST__   | `/api/url/_redirect`                | Redirects the user using HTTP 302 (temporary) redirect. Receives URL generator ID and parameters in the POST body payload.               |
```

Comment:
do we really want to 302, or should we use new platforms navigateWithState ? seems with the latter we will be able to navigate into apps that don't expose full state in the url without them providing special endpoint just for that case.

Reply:
++ Yeah I think this is a key consideration, IMO the primary benefit of a redirect route is that we can use it to proxy state to apps which otherwise wouldn't support state in the url.

As long as apps are already handling state from `navigateToApp`, then they wouldn't need to do anything differently... the redirect route would just take the state and an app ID, and call `navigateWithState` on the client. (You lose the nice semantics of a proper 302 redirect however)

Reply:
My thinking was that currently Cloud has a lot of deep links into Dashboad and Discover and those links are sometimes breaking. They could use a redirect route with URL generator to redirect them to the correct link, and it would use the proper 302 redirect semantics as look mentions, and the browser/proxy would cache that redirect.

So external use like Cloud could use the server redirect endpoint because:

1. The current Cloud linking use case does not need to send location state to the app.
2. Nice 302 semantics.
3. Possibly more efficient, as to execute the `navigateToApp` we need to load Kibana client-side, but to redirect user or just construct a URL, when we don't need the location state, can be done without loading Kibana client-side.

But, of course, we can remove this endpoint and just use a browser-side equivalent.

Reply:
Discussing it with Peter today agreed to only have a client-side redirect for smaller API surface. Like so:

```
/app/goto/_redirect/DISCOVER_DEEP_LINKS?params={"indexPattern":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx","highlightedField":"foo"}&paramsVersion=7.x
```